### PR TITLE
add ray quickstart

### PR DESCRIPTION
**Pull request by yaxinchen666**

### CODE REVIEW

File: docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md [hunk 1]
```diff
@@ -0,0 +1,194 @@
+# Ray Quickstart
+
+---
+
+![](../../../../image/colab_logo_32px.png)[Run in Google Colab](https://colab.research.google.com/github/intel-analytics/analytics-zoo/blob/master/docs/docs/colab-notebook/orca/quickstart/ray_sharded_parameter_server.ipynb) &nbsp;![](../../../../image/GitHub-Mark-32px.png)[View source on GitHub](https://github.com/intel-analytics/analytics-zoo/blob/master/docs/docs/colab-notebook/orca/quickstart/ray_sharded_parameter_server.ipynb)
+
+---
+
+**In this guide we will describe how to run Ray applications using the RayOnSpark support in Analytics Zoo in 2 simple steps.**
```
Comment (hkvision) [inline docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md:9]: run Ray applications on big data clusters using RayOnSpark in 2 simple steps.

File: docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+conda activate zoo
+pip install analytics-zoo[ray]
+```
+
+### **Step 1: Initialize**
+
+We recommend using `init_orca_context` to initiate and run Analytics Zoo on the underlying cluster. The Ray cluster would be launched as well by specifying `init_ray_on_spark=True`.
+
+```python
+if args.cluster_mode == "local":
```
Comment (hkvision) [inline docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md:26]: Remove args.cluster mode, change to comment:
# For local machine
# For K8s cluster
# For Hadoop/YARN cluster

File: docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    sc = init_orca_context(cluster_mode="local", cores=4, init_ray_on_spark=True)# run in local mode
+elif args.cluster_mode == "k8s":
+    sc = init_orca_context(cluster_mode="k8s", num_nodes=2, cores=2, init_ray_on_spark=True) # run on K8s cluster
+elif args.cluster_mode == "yarn":
+    sc = init_orca_context(cluster_mode="yarn-client", cores=2, memory="10g", num_nodes=2, init_ray_on_spark=True) # run on Hadoop YARN cluster
+```
+
```
Comment (hkvision) [inline docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md:33]: missing this sentence?
```
This is the only place where you need to specify local or distributed mode. View Orca Context for more details.
```

File: docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md (hunk 2 above)
Comment (hkvision) [reply docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md:26]: add the import sentence

File: docs/docs/colab-notebook/orca/quickstart/ray_sharded_parameter_server.ipynb [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+    "!pip install ray==1.2.0"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "oMXB8SmUPtvr"
+   },
+   "source": [
+    "## **Run Ray applications using Analytics Zoo**\n",
```
Comment (jason-dai) [inline docs/docs/colab-notebook/orca/quickstart/ray_sharded_parameter_server.ipynb:135]: RayOnSpark: run Ray programs directly on Big Data clusters

File: docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+We recommend using `init_orca_context` to initiate and run Analytics Zoo on the underlying cluster. The Ray cluster would be launched as well by specifying `init_ray_on_spark=True`.
+
+```python
+if args.cluster_mode == "local":
+    sc = init_orca_context(cluster_mode="local", cores=4, init_ray_on_spark=True)# run in local mode
+elif args.cluster_mode == "k8s":
+    sc = init_orca_context(cluster_mode="k8s", num_nodes=2, cores=2, init_ray_on_spark=True) # run on K8s cluster
+elif args.cluster_mode == "yarn":
+    sc = init_orca_context(cluster_mode="yarn-client", cores=2, memory="10g", num_nodes=2, init_ray_on_spark=True) # run on Hadoop YARN cluster
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md:31]: cluster_mode="yarn"

Why the parameters for yarn and k8s are different?

Comment (hkvision) [reply docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md:31]: Make them the same @yaxinchen666 specify cores, memory and num_nodes for both.

File: docs/docs/colab-notebook/orca/quickstart/ray_sharded_parameter_server.ipynb [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+    "\n",
+    "In this guide we will describe how to run Ray applications using the RayOnSpark support in Analytics Zoo in 2 simple steps."
```
Comment (jason-dai) [inline docs/docs/colab-notebook/orca/quickstart/ray_sharded_parameter_server.ipynb:137]: In this guide, we will describe how to use RayOnSpark to directly run Ray programs on Big Data clusters in 2 simple steps.

File: docs/docs/colab-notebook/orca/quickstart/ray_sharded_parameter_server.ipynb [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "YsTyFBA10OVJ"
+   },
+   "source": [
+    "### **Step 1: Initialize**\n",
+    "We recommend using `init_orca_context` to initiate and run Analytics Zoo on the underlying cluster. The Ray cluster would be launched as well by specifying `init_ray_on_spark=True`."
```
Comment (jason-dai) [inline docs/docs/colab-notebook/orca/quickstart/ray_sharded_parameter_server.ipynb:161]: The Ray cluster would be launched automatically by specifying `init_ray_on_spark=True`

File: docs/docs/colab-notebook/orca/quickstart/ray_sharded_parameter_server.ipynb [hunk 8]
```diff
@@ -0,0 +1,451 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "YsTyFBA10OVJ"
+   },
+   "source": [
+    "### **Step 2: Run Ray Applications**\n",
+    "After the initialization, you can directly run Ray applications on the underlying cluster. Ray [tasks](https://docs.ray.io/en/master/walkthrough.html#remote-functions-tasks) and [actors](https://docs.ray.io/en/master/actors.html) would be launched across the cluster.\n",
```
Comment (jason-dai) [inline docs/docs/colab-notebook/orca/quickstart/ray_sharded_parameter_server.ipynb:209]: After the initialization, you can directly write Ray code inline with your Spark code, and run Ray programs on the underlying existing Big Data clusters.

File: docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md (hunk 5 above)
Comment (yaxinchen666) [reply docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md]: Sorry that I referred to two different guides. Fixed.

**Comment by yaxinchen666** [comment]
Thanks for comments. Fixed.

File: docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+```python
+from zoo.orca import init_orca_context
+
+# For local machine
+sc = init_orca_context(cluster_mode="local", cores=4, memory="10g", init_ray_on_spark=True)
+# For K8s cluster
+sc = init_orca_context(cluster_mode="k8s", num_nodes=2, cores=2, memory="10g", init_ray_on_spark=True)
+# For Hadoop/YARN cluster
+sc = init_orca_context(cluster_mode="yarn-client", num_nodes=2, cores=2, memory="10g", init_ray_on_spark=True)
```
Comment (hkvision) [inline docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md:33]: change yarn-client to yarn

File: docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+### **Step 1: Initialize**
+
+We recommend using `init_orca_context` to initiate and run Analytics Zoo on the underlying cluster. The Ray cluster would be launched automatically by specifying `init_ray_on_spark=True`.
+
+```python
+from zoo.orca import init_orca_context
+
+# For local machine
+sc = init_orca_context(cluster_mode="local", cores=4, memory="10g", init_ray_on_spark=True)
```
Comment (hkvision) [inline docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md:29]: cores=2 to align with others?

File: docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md (hunk 2 above)
Comment (jason-dai) [reply docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md]: > Remove args.cluster mode, change to comment:
> 
> ```
> # For local machine
> # For K8s cluster
> # For Hadoop/YARN cluster
> ```

I think we can keep it (as consistent with other quick start)

File: docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md (hunk 10 above)
Comment (yaxinchen666) [reply docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md:29]: I checked other quickstart guides and I think cores in local = num_nodes * cores in distributed?

File: docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md (hunk 2 above)
Comment (hkvision) [reply docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md:28]: What about modifying the other quick starts as well? Really weird to see args...

File: docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md (hunk 10 above)
Comment (hkvision) [reply docs/readthedocs/source/doc/Orca/QuickStart/ray-quickstart.md]: okay, it makes sense

File: docs/docs/colab-notebook/orca/quickstart/ray_sharded_parameter_server.ipynb [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "0Si5BdI7yJ2A"
+   },
+   "source": [
+    "**Install Analytics Zoo**\n",
+    "\n",
+    "You can install the latest pre-release version using `!pip install --pre  analytics-zoo`. "
```
Comment (jason-dai) [inline docs/docs/colab-notebook/orca/quickstart/ray_sharded_parameter_server.ipynb:101]: shall we use `!pip install --pre --upgrade analytics-zoo` as described at https://analytics-zoo.readthedocs.io/en/latest/doc/UserGuide/python.html @hkvision?

Comment (hkvision) [reply docs/docs/colab-notebook/orca/quickstart/ray_sharded_parameter_server.ipynb:101]: yeah, need to update this @yaxinchen666

Comment (yaxinchen666) [reply docs/docs/colab-notebook/orca/quickstart/ray_sharded_parameter_server.ipynb]: Done.

**Comment by hkvision** [comment]
Move the notebook from orca/quickstart to ray (create a new folder)?
Also need to add test scripts in .sh: https://github.com/intel-analytics/analytics-zoo/tree/master/docs/docs/colab-notebook
Probably create a new sh file? Or can reuse tf2.sh?

**Comment by yaxinchen666** [comment]
> Move the notebook from orca/quickstart to ray (create a new folder)?
I moved it to ray/quickstart. Seems ray folder already exists because of RayUserGuide.
> Also need to add test scripts in .sh: https://github.com/intel-analytics/analytics-zoo/tree/master/docs/docs/colab-notebook
> Probably create a new sh file? Or can reuse tf2.sh?
I created a run-notebook-ray.sh. The content is almost the same with tf2.sh.

**Comment by hkvision** [comment]
http://10.239.47.210:18888/job/ZOO-PR-Quickstart-notebook/ Can we create a job for Colab-Ray? @Le-Zheng

File: docs/readthedocs/source/doc/Ray/QuickStart/ray-quickstart.md [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+### **Step 1: Initialize**
+
+We recommend using `init_orca_context` to initiate and run Analytics Zoo on the underlying cluster. The Ray cluster would be launched automatically by specifying `init_ray_on_spark=True`.
+
+```python
+from zoo.orca import init_orca_context
+
+if cluster_mode == "local":
+    # For local machine
```
Comment (hkvision) [inline docs/readthedocs/source/doc/Ray/QuickStart/ray-quickstart.md:29]: put this comment in the same line with if
```
if cluster_mode == "local": # For local machine
```

File: docs/readthedocs/source/doc/Ray/QuickStart/ray-quickstart.md [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+    sc = init_orca_context(cluster_mode="local", cores=4, memory="10g", init_ray_on_spark=True)
+elif cluster_mode == "k8s":
+    # For K8s cluster
+    sc = init_orca_context(cluster_mode="k8s", num_nodes=2, cores=2, memory="10g", init_ray_on_spark=True)
+elif cluster_mode == "yarn":
+    # For Hadoop/YARN cluster
+    sc = init_orca_context(cluster_mode="yarn", num_nodes=2, cores=2, memory="10g", init_ray_on_spark=True)
+```
+
+This is the only place where you need to specify local or distributed mode. View [Orca Context](./../Overview/orca-context.md) for more details.
```
Comment (hkvision) [inline docs/readthedocs/source/doc/Ray/QuickStart/ray-quickstart.md:39]: The link is invalid...

File: docs/readthedocs/source/doc/Ray/QuickStart/ray-quickstart.md [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+ray_ctx = OrcaContext.get_ray_context()
+address_info = ray_ctx.address_info  # The dictionary information of the ray cluster, including node_ip_address, object_store_address, webui_url, etc.
+redis_address = ray_ctx.redis_address  # The redis address of the ray cluster.
+```
+
+### **Step 2: Run Ray Applications**
+
+After the initialization, you can directly write Ray code inline with your Spark code, and run Ray programs on the underlying existing Big Data clusters. Ray [tasks](https://docs.ray.io/en/master/walkthrough.html#remote-functions-tasks) and [actors](https://docs.ray.io/en/master/actors.html) would be launched across the cluster.
+
+The following example uses actor handles to implement a parameter server example for distributed asynchronous stochastic gradient descent.
```
Comment (hkvision) [inline docs/readthedocs/source/doc/Ray/QuickStart/ray-quickstart.md:57]: Add a sentence like `This is a simple Ray example for demonstration purpose. Similarly, you can write other Ray applications as you wish.`?

File: docs/readthedocs/source/doc/Ray/QuickStart/ray-quickstart.md (hunk 13 above)
Comment (yaxinchen666) [reply docs/readthedocs/source/doc/Ray/QuickStart/ray-quickstart.md]: Oooops sorry that I forgot that I changed its location. Fixed.

File: docs/readthedocs/source/doc/Ray/QuickStart/ray-quickstart.md (hunk 14 above)
Comment (yaxinchen666) [reply docs/readthedocs/source/doc/Ray/QuickStart/ray-quickstart.md]: Cool.

**Comment by Le-Zheng** [comment]
> http://10.239.47.210:18888/job/ZOO-PR-Quickstart-notebook/ Can we create a job for Colab-Ray? @Le-Zheng

Job created: http://10.239.47.210:18888/view/ZOO-PR/job/ZOO-PR-Colab-ray/

**Comment by hkvision** [comment]
> > http://10.239.47.210:18888/job/ZOO-PR-Quickstart-notebook/ Can we create a job for Colab-Ray? @Le-Zheng
> 
> Job created: http://10.239.47.210:18888/view/ZOO-PR/job/ZOO-PR-Colab-ray/

Thanks. We would test on this job.

**Comment by yaxinchen666** [comment]
> > http://10.239.47.210:18888/job/ZOO-PR-Quickstart-notebook/ Can we create a job for Colab-Ray? @Le-Zheng
> 
> Job created: http://10.239.47.210:18888/view/ZOO-PR/job/ZOO-PR-Colab-ray/

Thanks. Jenkins passed: http://10.239.47.210:18888/job/ZOO-PR-Colab-ray/2/

**Comment by hkvision** [comment]
Nice result: https://rayquickstart4.readthedocs.io/en/latest/doc/Ray/QuickStart/ray-quickstart.html

File: docs/docs/colab-notebook/ray/quickstart/ray_sharded_parameter_server.ipynb [hunk 15]
```diff
@@ -0,0 +1,338 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "view-in-github"
+   },
+   "source": [
+    "<a href=\"https://colab.research.google.com/github/intel-analytics/analytics-zoo/blob/master/docs/docs/colab-notebook/orca/quickstart/keras_lenet_mnist.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
```
Comment (hkvision) [inline docs/docs/colab-notebook/ray/quickstart/ray_sharded_parameter_server.ipynb:10]: why there is keras_lenet_mnist?

**Comment by hkvision** [comment]
What about changing ray_sharded_parameter_server.ipynb to parameter_server.ipynb? Since the sharded part has been removed. Don't forget to change the name in the link of the quick start.

File: docs/docs/colab-notebook/ray/quickstart/ray_sharded_parameter_server.ipynb [hunk 16]
```diff
@@ -0,0 +1,338 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "qAWBOUouPj8S"
+   },
+   "outputs": [],
+   "source": [
+    "# Install python dependencies\n",
+    "!pip install ray==1.2.0"
```
Comment (hkvision) [inline docs/docs/colab-notebook/ray/quickstart/ray_sharded_parameter_server.ipynb:126]: remove this and add to analytics-zoo[ray]

File: docs/docs/colab-notebook/ray/quickstart/ray_sharded_parameter_server.ipynb [hunk 17]
```diff
@@ -0,0 +1,338 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "ray_ctx = OrcaContext.get_ray_context()\n",
+    "address_info = ray_ctx.address_info  # The dictionary information of the ray cluster, including node_ip_address, object_store_address, webui_url, etc.\n",
+    "redis_address = ray_ctx.redis_address  # The redis address of the ray cluster."
```
Comment (hkvision) [inline docs/docs/colab-notebook/ray/quickstart/ray_sharded_parameter_server.ipynb:200]: print address_info here?

File: docs/docs/colab-notebook/ray/quickstart/ray_sharded_parameter_server.ipynb [hunk 18]
```diff
@@ -0,0 +1,338 @@
...
+    "As the worker tasks are executing, you can query the parameter server from the driver and see the parameters changing in the background."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "print(ray.get(ps.get_parameters.remote()))"
```
Comment (hkvision) [inline docs/docs/colab-notebook/ray/quickstart/ray_sharded_parameter_server.ipynb:291]: why there is not output of this cell?

Comment (yaxinchen666) [reply docs/docs/colab-notebook/ray/quickstart/ray_sharded_parameter_server.ipynb:291]: I clear all of the outputs. Should I leave them there or just what is printed?

File: docs/docs/colab-notebook/ray/quickstart/ray_sharded_parameter_server.ipynb (hunk 15 above)
Comment (yaxinchen666) [reply docs/docs/colab-notebook/ray/quickstart/ray_sharded_parameter_server.ipynb:10]: Sorry it is originally copied from keras_lenet_mnist... Will modify it.

File: docs/docs/colab-notebook/ray/quickstart/ray_sharded_parameter_server.ipynb (hunk 18 above)
Comment (hkvision) [reply docs/docs/colab-notebook/ray/quickstart/ray_sharded_parameter_server.ipynb:291]: It's better to leave this so that viewers can know the weight?

Comment (yaxinchen666) [reply docs/docs/colab-notebook/ray/quickstart/ray_sharded_parameter_server.ipynb:291]: ok

File: docs/docs/colab-notebook/run-notebook-ray.sh [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+	rm ${ANALYTICS_ZOO_HOME}/docs/docs/colab-notebook/$2.py
+}
+
+# the first argument is the number of ipynb, the second argument is the name of ipynb,
+# the third argument is the runtime used by this notebook
+echo_time(){
+	echo "#$1 $2 time used: $3 seconds"
+}
+
+name1="ray/quickstart/ray_sharded_parameter_server"
```
Comment (hkvision) [inline docs/docs/colab-notebook/run-notebook-ray.sh:40]: this name needs to be changed as well

Comment (yaxinchen666) [reply docs/docs/colab-notebook/run-notebook-ray.sh]: Fixed. Should I run jenkins again?

File: docs/docs/colab-notebook/ray/quickstart/ray_parameter_server.ipynb [hunk 20]
```diff
@@ -0,0 +1,344 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "{'node_ip_address': '10.239.44.107', 'raylet_ip_address': '10.239.44.107', 'redis_address': '10.239.44.107:6379', 'object_store_address': '/tmp/ray/session_2021-03-30_13-56-46_494121_21728/sockets/plasma_store', 'raylet_socket_name': '/tmp/ray/session_2021-03-30_13-56-46_494121_21728/sockets/raylet', 'webui_url': '10.239.44.107:8265', 'session_dir': '/tmp/ray/session_2021-03-30_13-56-46_494121_21728', 'metrics_export_port': 48952, 'node_id': '3528c02f53bc6fb5db102383a2dcaa0bd65bedfea798434f0e56a30b'}\n"
```
Comment (hkvision) [inline docs/docs/colab-notebook/ray/quickstart/ray_parameter_server.ipynb:189]: print out only object_store_address

**Comment by yaxinchen666** [comment]
Jenkins passed: http://10.239.47.210:18888/view/ZOO-PR/job/ZOO-PR-Colab-ray/3/

**Review by hkvision — APPROVED** [review]
LGTM

**Pull request merged**